subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskCollection.java

```diff
@@ -95,22 +96,22 @@ protected UnknownTaskException createNotFoundException(String name) {
     }
 
     @Nullable
-    private ProviderInternal<? extends Task> findTask(String name) {
+    ProviderInternal<? extends Task> findTask(String name) {
         Task task = findByNameWithoutRules(name);
         if (task == null) {
             return findByNameLaterWithoutRules(name);
         }
-        Class<T> type = (Class<T>) getType();
-        return new TaskLookupProvider(type, name);
+        return Cast.uncheckedCast(getInstantiator().newInstance(TaskLookupProvider.class, this, getType(), name));
     }
 
-
-    protected abstract class DefaultTaskProvider<T extends Task> extends AbstractProvider<T> implements Named, TaskProvider<T> {
+    public static abstract class DefaultTaskProvider<T extends Task> extends AbstractProvider<T> implements Named, TaskProvider<T> {
```

Comment:
we should mark `TaskProvider` as `@NonExtensible` so that only the DSL stuff is mixed in and none of the extensibility stuff is mixed in. I don't think we want to make the providers extensible, as it doesn't really make any sense in this case and it adds some performance cost.